addon/adapters/firebase.js

```diff
@@ -0,0 +1,516 @@
+    for saving nested records as well.
+
+  */
+  updateRecord: function(store, type, record, _recordRef) {
+    var adapter = this;
+    var recordRef = _recordRef || this._getRef(type, record.id);
+    var recordCache = Ember.get(adapter._recordCacheForType, fmt('%@.%@', [type.typeKey, record.get('id')])) || {};
+
+    var serializedRecord = record.serialize({includeId:false});
+
+    return new Promise(function(resolve, reject) {
+      var savedRelationships = Ember.A();
+      record.eachRelationship(function(key, relationship) {
+        var save;
+        if (relationship.kind === 'hasMany') {
+            if (serializedRecord[key]) {
+              save = adapter._saveHasManyRelationship(store, type, relationship, serializedRecord[key], recordRef, recordCache);
+              savedRelationships.push(save);
+              // Remove the relationship from the serializedRecord because otherwise we would clobber the entire hasMany
+              delete serializedRecord[key];
+            }
+        } else {
+            if (relationship.options.embedded === true && serializedRecord[key]) {
+              save = adapter._saveBelongsToRecord(store, type, relationship, serializedRecord[key], recordRef);
+              savedRelationships.push(save);
+              delete serializedRecord[key];
+            }
+        }
+      });
+
+      var relationshipsPromise = Ember.RSVP.allSettled(savedRelationships);
+      var recordPromise = adapter._updateRecord(recordRef, serializedRecord);
+
+      Ember.RSVP.hashSettled({relationships: relationshipsPromise, record: recordPromise}).then(function(promises) {
+        var rejected = Ember.A(promises.relationships.value).filterBy('state', 'rejected');
+        if (promises.record.state === 'rejected') {
+          rejected.push(promises.record);
+        }
+        // Throw an error if any of the relationships failed to save
+        if (rejected.length !== 0) {
+          var error = new Error(fmt('Some errors were encountered while saving %@ %@', [type, record.id]));
+              error.errors = rejected.mapBy('reason');
+          reject(error);
+        } else {
+          resolve();
+        }
+      });
+    }, fmt('DS: FirebaseAdapter#updateRecord %@ to %@', [type, recordRef.toString()]));
+  },
+
+  //Just update the record itself without caring for the relationships
+  _updateRecord: function(recordRef, serializedRecord) {
+    return toPromise(recordRef.update, recordRef, [serializedRecord]);
+  },
+
+  /**
+    Call _saveHasManyRelationshipRecord on each record in the relationship
+    and then resolve once they have all settled
+  */
+  _saveHasManyRelationship: function(store, type, relationship, ids, recordRef, recordCache) {
+    if (!Ember.isArray(ids)) {
+      throw new Error('hasMany relationships must must be an array');
+    }
+    var adapter = this;
+    var idsCache = Ember.A(recordCache[relationship.key]);
+    ids = Ember.A(ids);
```

Comment:
due to the size of the refactor, I don't want to do much alteration to the adapter/serializer code here. I'd prefer to do the next stage of alteration once we have this merged into master. Seems like  Matt created this change as a new PR anyway.